### PR TITLE
ICU-22498 Write directly to sink (instead of into legacy buffers)

### PR DESCRIPTION
**Pull request by roubert**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22498
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Review by markusicu — APPROVED** [review]
lgtm tnx

File: icu4c/source/common/uloc_tag.cpp [hunk 1]
```diff
@@ -1945,37 +1942,18 @@ _appendPrivateuseToLanguageTag(const char* localeID, icu::ByteSink& sink, UBool
...
-                            if (reslen < capacity) {
-                                uprv_memcpy(tmpAppend + reslen, PRIVUSE_VARIANT_PREFIX, uprv_min(len, capacity - reslen));
-                                reslen += uprv_min(len, capacity - reslen);
-                            }
-
-                            if (reslen < capacity) {
-                                tmpAppend[reslen++] = SEP;
-                            }
-
+                            sink.Append(PRIVATEUSE_KEY, UPRV_LENGTHOF(PRIVATEUSE_KEY) - 1);
```
Comment (FrankYFTang) [inline icu4c/source/common/uloc_tag.cpp:1948]: why "- 1" here?

Comment (FrankYFTang) [inline icu4c/source/common/uloc_tag.cpp:1948]: Is  uprv_strlen(PRIVATEUSE_KEY) better?

File: icu4c/source/common/uloc_tag.cpp [hunk 2]
```diff
@@ -1945,37 +1942,18 @@ _appendPrivateuseToLanguageTag(const char* localeID, icu::ByteSink& sink, UBool
...
+                            sink.Append("-", 1);
+                            sink.Append(PRIVUSE_VARIANT_PREFIX, UPRV_LENGTHOF(PRIVUSE_VARIANT_PREFIX) - 1);
+                            sink.Append("-", 1);
```
Comment (FrankYFTang) [inline icu4c/source/common/uloc_tag.cpp:1951]: Is  uprv_strlen(PRIVUSE_VARIANT_PREFIX) better?

File: icu4c/source/common/uloc_tag.cpp (hunk 1 above)
Comment (roubert) [reply icu4c/source/common/uloc_tag.cpp:1948]: I would find it a bit odd to call a function to dynamically calculate the length of a static string constant that has a length known at compile time.

File: icu4c/source/common/uloc_tag.cpp (hunk 2 above)
Comment (roubert) [reply icu4c/source/common/uloc_tag.cpp:1951]: I would find it a bit odd to call a function to dynamically calculate the length of a static string constant that has a length known at compile time.

**Pull request merged**